File: code/packages/fsharp/hash-set/HashSet.fs
namespace CodingAdventures.HashSet.FSharp

open CodingAdventures.HashMap.FSharp

type HashSet<'T when 'T : equality>(?entries: seq<'T>) =
    let map =
        entries
        |> Option.defaultValue Seq.empty
        |> Seq.fold (fun (state: HashMap<'T, bool>) value -> state.Set(value, true)) (HashMap<'T, bool>())

    member _.Count = map.Size
    member _.Size = map.Size

    member _.Clone() =
        HashSet<'T>(map.Keys())

    member _.Has(value: 'T) = map.Has value
    member _.Contains(value: 'T) = map.Has value
    member _.IsEmpty() = map.Size = 0
    member _.ToList() = map.Keys()

    member this.Add(value: 'T) =
        HashSet<'T>((value :: this.ToList()) |> Seq.distinct)

    member this.Remove(value: 'T) =
        HashSet<'T>(this.ToList() |> List.filter ((<>) value))

    member this.Union(other: HashSet<'T>) =
        HashSet<'T>(Seq.append (this.ToList()) (other.ToList()) |> Seq.distinct)

    member this.Intersection(other: HashSet<'T>) =
        HashSet<'T>(this.ToList() |> List.filter other.Has)

    member this.Difference(other: HashSet<'T>) =
        HashSet<'T>(this.ToList() |> List.filter (other.Has >> not))

File: code/packages/haskell/hash-set/src/HashSet.hs
module HashSet
    ( description
    , HashSet
    , empty
    , singleton
    , isEmpty
    , size
    , contains
    , add
    , remove
    , fromList
    , toList
    , union
    , intersection
    , difference
    ) where

import qualified Data.Set as Set

description :: String
description = "Haskell hash set for DT19 and Mini Redis"

newtype HashSet value = HashSet
    { unHashSet :: Set.Set value
    }
    deriving (Eq, Show)

empty :: HashSet value
empty = HashSet Set.empty

singleton :: Ord value => value -> HashSet value
singleton value = HashSet (Set.singleton value)

isEmpty :: HashSet value -> Bool
isEmpty (HashSet valuesSet) = Set.null valuesSet

size :: HashSet value -> Int
size (HashSet valuesSet) = Set.size valuesSet

contains :: Ord value => value -> HashSet value -> Bool
contains value (HashSet valuesSet) = Set.member value valuesSet

add :: Ord value => value -> HashSet value -> HashSet value
add value (HashSet valuesSet) = HashSet (Set.insert value valuesSet)

remove :: Ord value => value -> HashSet value -> HashSet value
remove value (HashSet valuesSet) = HashSet (Set.delete value valuesSet)

fromList :: Ord value => [value] -> HashSet value
fromList = HashSet . Set.fromList

toList :: HashSet value -> [value]
toList (HashSet valuesSet) = Set.toAscList valuesSet

union :: Ord value => HashSet value -> HashSet value -> HashSet value
union (HashSet leftSet) (HashSet rightSet) = HashSet (Set.union leftSet rightSet)

intersection :: Ord value => HashSet value -> HashSet value -> HashSet value
intersection (HashSet leftSet) (HashSet rightSet) =
    HashSet (Set.intersection leftSet rightSet)

difference :: Ord value => HashSet value -> HashSet value -> HashSet value
difference (HashSet leftSet) (HashSet rightSet) =
    HashSet (Set.difference leftSet rightSet)
